### PR TITLE
feat(VDatePicker): add date range first & last classes

### LINKED ISSUE #12851: [Feature Request] Different classes for first in range and last in range in VDatePicker ranged
### Problem to solve
Custom layout for range pickers

### Proposed solution
add range-start class to the first day in range and add range-end class to the last day in range

### PR DESCRIPTION
**Pull request by joelpro2**
## Description

Added class v-date-picker--first_in_range to the first day in range
Added class v-date-picker--last_in_range to the last day in range
Added test for both

resolves #12851 

## Motivation and Context
This change allows devs to style the threshold days
 #12851 

## How Has This Been Tested?
unit

## Markup:

<details>

```vue
// Paste your FULL Playground.vue here
<template>
  <v-container>
    <v-date-picker
      v-model="dates"
      range
    ></v-date-picker>
  </v-container>
</template>

<script>
  export default {
    data: () => ({
      dates: []
    }),
  }
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Review by jacekkarczmarczyk — CHANGES_REQUESTED** [review]
1. please rebase properly, pr contains some unrelated commits
2. we don't use capital letters in class names
3. it applies classes to the button, i think it's better if it applied classes to the `<td>`
4. afair dates in v-model may be sorted descending if you click the last date of a range as first (didn't check it though)
5. don't you need these classes also for month/year selector?

**Comment by joelpro2** [comment]
> 2. we don't use capital letters in class names -> How would you call it then?
> 3. it applies classes to the button, i think it's better if it applied classes to the `<td>` -> i'm applying classes where you are
> 4. afair dates in v-model may be sorted descending if you click the last date of a range as first (didn't check it though) ->this is for people to decide, in my project i sort the array but i think that may be cases that the beggining of the range may not be the oldest date.
> 5. don't you need these classes also for month/year selector? -> Maybe for the month.

**Comment by jacekkarczmarczyk** [comment]
> > 1. we don't use capital letters in class names.
> 
> How would you call it then?

kebab case:
`some-block__some-element--some-modifier`

> > 1. it applies classes to the button, i think it's better if it applied classes to the `<td>`
> 
> I'm applying classes where you are

I know but i think that applying it to `<td>` would be more flexible for styling

> > 1. afair dates in v-model may be sorted descending if you click the last date of a range as first (didn't check it though)
> 
> This is for people to decide, in my project i sort the array but i think that may be cases that the beggining of the range may not be the oldest date.

https://codepen.io/jkarczm/pen/xxEYZNd - select 31dec then 1dec and you'll end up with this
image

**Comment by joelpro2** [comment]
Is this still missing something? @MajesticPotatoe @jacekkarczmarczyk

**Comment by jacekkarczmarczyk** [comment]
@joelpro2 patience please ;) it's planned for 2.5 which is not planned for the next few days

Looks like you're still missing applying the classes based on the order of dates in the array instead of order of the dates, so if the model contains `['2000-01-11', '2000-01-01']` the `2000-01-11` will be considered as first and `2000-01-01` as last, which imho isn't very helpful in terms of styling

**Comment by KaelWD** [comment]
@joelpro2 are you still working on this?

**Comment by joelpro2** [comment]
> @joelpro2 are you still working on this?

No, like i said, the dates order should be up to developers to decide wherever the want dates ordered or not, and for the classes, it makes no sense to remove them from where the other styles are already applied, it would force developers to target more elements without a need.

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 1]
```diff
@@ -93,7 +93,8 @@ export default mixins(
   },
 
   methods: {
-    genButtonClasses (isAllowed: boolean, isFloating: boolean, isSelected: boolean, isCurrent: boolean) {
+    genButtonClasses (isAllowed: boolean, isFloating: boolean, isSelected: boolean, isCurrent: boolean,
+      isFirst: boolean, isLast: boolean) {
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:97]:
```suggestion
    genButtonClasses (
      isAllowed: boolean, 
      isFloating: boolean, 
      isSelected: boolean, 
      isCurrent: boolean,
      isFirst: boolean, 
      isLast: boolean,
     ) {
```

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 2]
```diff
@@ -103,6 +104,8 @@ export default mixins(
         'v-btn--rounded': isFloating,
         'v-btn--disabled': !isAllowed || this.disabled,
         'v-btn--outlined': isCurrent && !isSelected,
+        'v-date-picker--first_in_range': isFirst,
+        'v-date-picker--last_in_range': isLast,
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:108]:
```suggestion
        'v-date-picker--first-in-range': isFirst,
        'v-date-picker--last-in-range': isLast,
```

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 3]
```diff
@@ -121,10 +124,17 @@ export default mixins(
...
+      if (this.range && !!this.value && Array.isArray(this.value)) {
+        isFirst = value === this.value[0]
+        isLast = value === this.value[this.value.length - 1]
+      }
 
       return this.$createElement('button', setColor(color, {
         staticClass: 'v-btn',
-        class: this.genButtonClasses(isAllowed && !isOtherMonth, isFloating, isSelected, isCurrent),
+        class: this.genButtonClasses(isAllowed && !isOtherMonth, isFloating, isSelected, isCurrent,
+          isFirst, isLast),
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:137]:
```suggestion
        class: this.genButtonClasses(
          isAllowed && !isOtherMonth, 
          isFloating, 
          isSelected, 
          isCurrent,
          isFirst, 
          isLast,
        ),
```

**Review by johnleider — CHANGES_REQUESTED** [review]

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 4]
```diff
@@ -93,8 +93,14 @@ export default mixins(
...
   methods: {
-    genButtonClasses (isAllowed: boolean, isFloating: boolean, isSelected: boolean, isCurrent: boolean,
-      isFirst: boolean, isLast: boolean) {
+    genButtonClasses (
+      isAllowed: boolean,
+      isFloating: boolean,
+      isSelected: boolean,
+      isCurrent: boolean,
+      isFirst: boolean,
+      isLast: boolean
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:102]:
```suggestion
      isLast: boolean,
```

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 5]
```diff
@@ -133,8 +139,14 @@ export default mixins(
...
         staticClass: 'v-btn',
-        class: this.genButtonClasses(isAllowed && !isOtherMonth, isFloating, isSelected, isCurrent,
-          isFirst, isLast),
+        class: this.genButtonClasses(
+          isAllowed && !isOtherMonth,
+          isFloating,
+          isSelected,
+          isCurrent,
+          isFirst,
+          isLast
```
Comment (johnleider) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:148]:
```suggestion
          isLast,
```

**Review by johnleider — CHANGES_REQUESTED** [review]

**Comment by johnleider** [comment]
Pulse check. About to close this as stale.

**Comment by joelpro2** [comment]
> Pulse check. About to close this as stale.

sorry, thought it was ready to be merged

**Review by johnleider — APPROVED** [review]

**Comment by johnleider** [comment]
One last once over from @KaelWD and we are good to go.

**Review by KaelWD — APPROVED** [review]

**Review by jacekkarczmarczyk — APPROVED** [review]

**Pull request merged**